### ISSUE TITLE
The case for passes

### ISSUE THREAD

**Issue by kvark** [gpuweb/gpuweb#64]
We've been going back and forth on the need for render/compute/blit passes, and I'd like to systematize the arguments and suggest models we could navigate from.

In native APIs, Vulkan has render passes only, Metal requires all the work to be done in one of the blit/compute/render pass. And D3D12 does not have any passes yet.

Here are the major areas (I can think of) that could take advantage of a pass concept, and the ideas on how they might be relevant to WebGPU:

### Resource bindings inheritance - #24

In Metal, resource bindings are only valid with the scope of a pass, where in Vulkan/D3D12 they are scoped to command buffers. This allows Metal runtime/driver to allocate descriptors based on everything used in a pass. This isn't a benefit for WebGPU, given that we have explicit resource bind groups created by the user.

### Synchronization and validation - #27, #59

In Vulkan the memory barriers and layout transitions are explicit. In D3D12 the resource state transitions as well as UAV/aliasing barriers are mostly explicit, but there are [rules](https://docs.microsoft.com/en-us/windows/desktop/direct3d12/using-resource-barriers-to-synchronize-resource-states-in-direct3d-12#common-state-promotion) for automatic to/from conversion to the "common" state. In Metal, memory barriers are mostly implicit. They are inserted automatically at the render pass boundaries, around individual dispatch and transfer calls.

For WebGPU, we decided to not expose the memory barriers in the API, and thus we can insert the appropriate synchronization commands between passes. In order for this to be valid, we introduce the "Usage" (mutable) state of a resource and declare a requirement of each resource having a constant usage across a pass. However, we also decided to automatically synchronize individual dispatch (UAV barriers only) and transfer calls, much like Metal. It isn't a long stretch to go from here to a full synchronization (allowing layout changes, for example), in which case the construct of a compute and transfer pass stops being useful (in the context of memory barriers).

### Work scheduling - #22

Hardware queues on Vulkan and D3D12 are explicit. In Metal, the `MTLCommandQueue` is also exposed, but it doesn't have a direct relation to a hardware queue. Metal runtime figures out the dependencies between passes (which is sound, given that *all* the work is split into passes) and [reserves the right](https://developer.apple.com/videos/play/wwdc2018/607/) to actually execute them on different hardware queues (note: how much this is the case in reality is unknown to me).

In WebGPU we decided to not expose the hardware queues, thus I find the opportunity of automatically taking advantage of them to be quite strong to miss.

### Type safety

Each kind of a pass has a restricted set of operations that can be recorded. E.g. one can't `dispatch()` or bind compute resources in a `MTLBlitCommandEncoder` pass. This makes the API more usable, assists documentation, and is generally is a good practice for API design.

**Comment by Kangz** [comment]
Thanks for the investigation.

The resource binding inheritance does help the D3D12 backend in the same way it helps the Metal driver: we might need to rewrite descriptors in heaps per-pass (with caching to avoid doing it if we can obviously), having all descriptors at the ready would help with that.

For synchronization the concept of compute pass helps, because we know no image layout changes are needed inside the pass, only UAV barriers. This both helps make efficient implementations, and convey to the app that the only "hidden" cost inside a compute pass would be because of UAV barriers. It makes WebGPU more efficient because it reduces the granularity at which layout transitions need to be considered and inserted.

We decided to not expose queues in the MVP of WebGPU but we're still interested in potentially having them in v1.

**Comment by kvark** [comment]
Here is a solution that I believe to be the most sound of our alternatives.
Instead of `WebGPUCommandEncoder` that we have now, expose two different types of encoders/passes:
  1. `WebGPURenderPass` - user provides the framebuffer (or render targets) to create one.
      - users are only allowed to change the render states (bind pipelines, resource groups, change the scissor, etc) and issue draw calls inside this pass
      - no synchronization is done by implementation within such a pass
  2. `WebGPUWorkPass` - no arguments are needed to create one.
      - users are only allowed to issue copy commands, bind compute resource groups, and issue compute dispatches inside this pass
      - implementation is expected to synchronize everything inside

Why I think this is a good idea:
  - render passes are one of the heaviest concept in every native API, so encompassing them into a separate construct on our side makes sense and sets the right expectations for the users
  - clear separation of synchronized versus non-synchronized zones
  - less API concepts than having a command buffer object plus 3 different passes, while still providing reasonable type safety

**Comment by kainino0x** [comment]
Does `Queue.Submit` (or `Device.Submit` if there are no queues) then take a sequence of `(RenderPass or WorkPass)`?

Would, e.g., the Vulkan backend construct a `VkCommandBuffer` when the RenderPass/WorkPass is created, or not until they are submitted?

How might secondary command buffers work? Would we have secondary RenderPass and secondary WorkPass types? Or ARE the RenderPass/WorkPass objects equivalent to secondary command buffers that are inserted into one big command buffer at `Queue.Submit`?

**Comment by kvark** [comment]
@kainino0x 
> Does Queue.Submit (or Device.Submit if there are no queues) then take a sequence of (RenderPass or WorkPass)?

Yes. We could also have a base interface for those two (`WebGPUPass`).

> Would, e.g., the Vulkan backend construct a VkCommandBuffer when the RenderPass/WorkPass is created, or not until they are submitted?

Given that passes have isolated resource scopes, it would make sense for Vulkan to encode them into different `VkCommandBuffer` objects. Preferably, not at submission time.

> How might secondary command buffers work? Would we have secondary RenderPass and secondary WorkPass types? Or ARE the RenderPass/WorkPass objects equivalent to secondary command buffers that are inserted into one big command buffer at Queue.Submit?

This is a hard problem, it depends on many factors, including some ISV feedback on what they want to use the secondary command buffers for. For example, we could expose (1) a D3D12-like bundle that can be encoded and then executed inside of a render pass. Another direction is to expose (2) Metal-like parallel command encoders, i.e. a variant of the render pass start command that would give you a `Sequence<WebGPURenderPass>` (of a given length) that you encode independently (but are executed sequentially). Finally, if we introduce the sub-pass concept, we could have a (3) `WebGPURenderPass` object returned per sub-pass.

There are many options. I think we need (1) + (3) provided in the end, but there is a long way of discussions to it. I don't think that accepting this proposal would harm our future exploration of those options.

**Comment by Kangz** [comment]
@kvark correct me if I'm wrong but my understanding is that your proposal tries to reduce the number of concepts and corner-cases users of WebGPU will need to internalize, as well as improve type-safety (even if it's not a thing in JS). Tweaking your proposal a little bit (which makes it close-ish to Metal), what if we have the following passes:

 1. `WebGPURenderPass` like in your proposal
 1. `WebGPUComputePass` that can only contain setting state for compute work and dispatches
 1. `WebGPUOtherWorkPass` (though the name sucks ^^) that contains everything but render and compute stuff.

In term of cognitive overhead I think it has the following advantages:
 - `WebGPUOtherWorkPass` doesn't have state the developer has to worry about.
 - `WebGPUComputePass` doesn't have non-programmable operations intermixed that may or may not cause stalls or other expensive things the developer might not expect.
 - `WebGPURenderPass` and `WebGPUComputePass` contain the execution of programs on the GPU and the state that controls it. That's it. To make it more clear we could make them inherit from a `WebGPUProgrammablePass` that have pipeline, bindgroup and pushconstants functions. (+ queries eventually).
 - Both `WebGPURenderPass` and `WebGPUComputePass` are essentially synchronized with caveats for UAVs that are unsynchronized in render passes and unsynchronized inside a single dispatch. The same cognitive burden that is present in your proposal too.

The pass object approach you highlighted has advantages compared to putting all commands at the same level in a command buffer: type-safety and less run-time errors because you forgot to begin or end a pass. There's still a bunch of unknowns like Kai said but it looks like an interesting direction.

**Comment by kvark** [comment]
@Kangz I like the concept of `WebGPUProgrammablePass` that both compute and render derive from. Overall, your modification sounds feasible to me.

**Comment by devshgraphicsprogramming** [comment]
Not exposing GPU synchronisation and injecting it implicitly is madness.
You will stall every single deferred shading pipeline that has been created since 2016.

**Comment by devshgraphicsprogramming** [comment]
Moreover foregoing queues, completely bricks the following performance saviours:
1. Asynchronous Compute
2. Asynchronous DMA data transfers to the GPU (no large-world streamed webMMOs for you)
3. Explicit Multi-GPU

**Comment by kvark** [comment]
@devshgraphicsprogramming 
> You will stall every single deferred shading pipeline that has been created since 2016.

Can you elaborate?

> Moreover foregoing queues, completely bricks the following performance saviours:

I agree that proper support for multi-queue/gpu may be hammered by implicit synchronization, but it's still to be figured out, see https://github.com/gpuweb/gpuweb/pull/95

**Comment by devshgraphicsprogramming** [comment]
> For WebGPU, we decided to not expose the memory barriers in the API, and thus we can insert the appropriate synchronization commands between passes. 

Basically you can have mutiple queues so that compute happens asynchronously to graphics work, this is good because as soon as you fill your GBuffer with rasterization you can kick off the deferred shading, postprocessing, SSAO, FXAA, Tonemapping, etc. in compute shaders.

Both Nvidia and AMD support async compute under vulkan, which means that the compute work for the previous frame can overlap with the rasterization of the GBuffer for the next, as long as the ordering is not overconstrained by over-the-top synchronisation.

Furthermore you should expose Vulkan-like sub-passes and the subpass dependencies explicitly so that tiled GPU architectures (all mobile, some Intel + NVidia Maxwell) can avoid resolving the fast local eDRAM to main memory between sub-passes.
This is what actually makes Deferred Rendering feasible on mobile GPUs (you read GBuffer in deferred shading pass straight from tiler scratch memory, not from texture).

**Comment by devshgraphicsprogramming** [comment]
Best example of overlapping compute and graphics for the same frame is compute shader LoD selection and culling while the Z-prepass is running.

Z-prepass usually severely underutilizes the Programmable Shading Cores, so there is some spare power to do compute shader with.
Nvidia and AMD both support running two or more OpenCL kernels/Graphics Shaders on the same GPU core simultaneously (CUDA explicitly says it can "hyperthread" and keep 2 different GPU programs resident on the same SM).

**Comment by magcius** [comment]
As far as my understanding goes, NVIDIA does not support asynchronous queues nor dual-issuing compute and graphics work. In fact, the stall in swapping between compute and graphics is a well-known issue that has only been fixed, I believe, in Volta.

**Comment by kvark** [comment]
@magcius I checked vulkan HW DB, picked a random Geforce MX150 and I'm seeing three different queue families (graphics, compute, transfer): https://vulkan.gpuinfo.org/displayreport.php?id=4560#queuefamilies

What would be the point of exposing those families differently if they get serialized (causing the stall on work switch)?

**Comment by devshgraphicsprogramming** [comment]
Pascal would be all of your 1050+ GPUs

Also the article you linked to mentions previous GPUs not being to able to perform compute and graphics simultaneously.
That does not change the fact that transfer queue (DMA) has been happening asynchronously since Tesla/Fermi on Nvidia.

Lastly I would like to note that not being able to issue dual compute and graphics =/= being able to dual issue different graphics work or different compute work.
In the very link you've provided it explicitly mentions CUDA being able to do so.

**Comment by devshgraphicsprogramming** [comment]
> `WebGPUWorkPass` - no arguments are needed to create one.
> 
>     * users are only allowed to issue copy commands, bind compute resource groups, and issue compute dispatches inside this pass
> 
>     * implementation is expected to synchronize everything inside

In that light, synchronising between compute dispatches is a Bad Idea (TM)

**Comment by devshgraphicsprogramming** [comment]
> As far as my understanding goes, NVIDIA does not support asynchronous queues nor dual-issuing compute and graphics work. In fact, the stall in swapping between compute and graphics is a well-known issue that has only been fixed, I believe, in Volta.

In fact, the very blog post series you've linked to outlines why more than one command queue is necessary, even for the same type of work.
https://mynameismjp.wordpress.com/2018/06/17/breaking-down-barriers-part-3-multiple-command-processors/

**Comment by kvark** [comment]
@devshgraphicsprogramming you are providing good input, just please try not to derail this issue into multi-queue discussion. *Unless* you have a strong reason to state that multi-queue submission is going to be impossible with the decisions made here.

Synchronization is discussed in #27 and #59. Multi-queue is in #22 and #95. This issue is mostly about integrating pieces of previous discussions and research into a more concrete API model.

> In that light, synchronising between all/every compute dispatches is a Bad Idea (TM)

It's precisely the topic of #59 (or rather, in #59 I proposed *not* to do that). The current consensus of the group though is that not having dispatches synchronized would be too error prone for the user (and Metal does that for the same reason). So you can see how Web conflicts with high-performance graphics here, and we are trying to find the right trade-offs.

You can help (aside from commenting, thank you for that!) by explaining the use-cases that are important and that would suffer most from the decisions chosen at the moment (e.g. synchronization between dispatches), which would make us (the W3C CG) reconsider.

**Comment by devshgraphicsprogramming** [comment]
> Unless you have a strong reason to state that multi-queue submission is going to be impossible with the decisions made here. 

If you insert synchronisation between every compute dispatch then having more than one "command processor" is useless, you've already completely serialized your operations into a dependency graph that is just a straight line.
The performance increase talked about in https://mynameismjp.wordpress.com/2018/06/17/breaking-down-barriers-part-3-multiple-command-processors/ cannot be achieved.

Multi-queue submission won't be impossible, it will just be pointless, as the work on one queue will not be able to proceed without the work completing on the other.

>  The current consensus of the group though is that not having dispatches synchronized would be too error prone for the user (and Metal does that for the same reason).

This is a next-generation low-level Graphics API for the web, why should we assume the user is inept at parallel programming?

Vulkan and DXIL make no such assumptions and give the developer the right tools to express the scheduling dependencies.
Maybe the creators of Metal don't hold their end-users in the same high regard?

My remark aside, if you indeed consider the approach to be too error prone, I suggest either:
1.  a scheduling layer/extension that can be applied on top to build a render(work)-graph of execution dependencies and help the user get the synchronisation right (submit the command buffers to the right queues and with the right sync primitives).
Many frameworks and engines are starting to do so, and DICE has a whole presentation about it.
2. If the user cannot grasp the parallelism and explicit synchronisation, they should not be using a low-level API... a middleware such as Unity, Unreal, Oryol, Sokol, The Forge, BGFX (if ported) and so on should solve that issue for them.

**Comment by kvark** [comment]
>If you insert synchronisation between every compute dispatch then having more than one "command processor" is useless, you've already completely serialized your operations into a dependency graph that is just a straight line.

I think you misunderstand what the "synchronization" means in this context. We aren't flushing the GPU pipeline after every dispatch, we are just making sure that the read/write and write/write accesses *to the same resource* are sequential. Hence, you can have multiple semi-independent dispatches in the same pass that are running in parallel. You can have multiple command queues doing parallel work, as long as they don't share access to a resource beyond just reading from it.

> This is a next-generation low-level Graphics API for the web, why should we assume the user is inept at parallel programming?

Short answer: Because the Web...

Long answer: we've tried to propose a lower level Vulkan-like API in https://github.com/KhronosGroup/WebGLNext-Proposals/tree/master/Obsidian-Mozilla , but it didn't get support from this group. Thus, if we want to ship anything (supported by all browsers), it's going to be a higher level API.

> If the user cannot grasp the parallelism and explicit synchronisation, they should not be using a low-level API

This isn't entirely false, but it's also rather harsh. Just because there isn't a good user-friendly API out there (aside from Metal, but I'm ok with marking this as arguable), doesn't mean that there couldn't be one. With new constraints from the Web environment, we are re-evaluating the decisions that native platforms did.

**Comment by Kangz** [comment]
@devshgraphicsprogramming this group is making a low-level API for the Web and one of the constraints of the Web is that applications should be portable by default meaning they produce the same result on all hardware / OS / browser combination. Vulkan isn't portable in that sense, for example if you forget synchronization your program might run fine on AMD GPUs but render garbage on NVIDIA GPUs.

> This is a next-generation low-level Graphics API for the web, why should we assume the user is inept at parallel programming?

We are not able to make WebGPU 100% deterministic but we should strive to. Barriers are too big of a footgun to give to applications and experience shows that very few people can get barriers right and there are Vulkan applications shipping today that are incorrect and just happen to work. Also I don't think WebGPU should be marketed as "low-level".

There was a lot of discussion on whether WebGPU should have implicit or explicit (validated) barriers. Unvalidated barriers were rejected early.

> If you insert synchronisation between every compute dispatch then having more than one "command processor" is useless, you've already completely serialized your operations into a dependency graph that is just a straight line.

The dependency graph is mostly serialized on each queue, but that doesn't prevent parallelism on multiple queues in itself. Muti-queue is discussed in #95 like @kvark pointed out.

> a scheduling layer/extension that can be applied on top to build a render(work)-graph of execution dependencies and help the user get the synchronisation right (submit the command buffers to the right queues and with the right sync primitives).

We thought about it but it is out of scope for now, both because we need the base of the API first, and because there isn't enough state of the art on this yet.

**Comment by devshgraphicsprogramming** [comment]
> we are just making sure that the read/write and write/write accesses to the same resource are sequential.

Why bother, the user can just as well write a poor compute shader that expects work-groups to synchronise between themselves.

You are worrying about making sure accesses of different dispatches stay sequential, when you will never be able to prevent work-groups within the same dispatch stay sequential.

> You can have multiple command queues doing parallel work, as long as they don't share access to a resource beyond just reading from it.

Again that will be a very big DAG, I think you're working under the presumption the DAG will represent a single-frame's worth of work.

This is not the case with higher-end applications where the DAG is never ending (overlapped frame processing and resource memory aliasing) and not very periodic (streaming).

> Vulkan isn't portable in that sense, for example if you forget synchronization your program might run fine on AMD GPUs but render garbage on NVIDIA GPUs.

AFAIK no API is portable according to that definition, except for consoles.
Also, no matter how good of a portable front-end API you design you will have to implement it in a back-end API, which is not portable.
With your current design directions you are setting yourself up to create a much more fleshed out API which requires a much more complex implementation leaving more room for non-portable behaviour and be far more at the mercy of the webGPU implementors.

> There was a lot of discussion on whether WebGPU should have implicit or explicit (validated) barriers. Unvalidated barriers were rejected early.

So explicit barriers were agreed upon or what?

**Comment by kvark** [comment]
>Why bother, the user can just as well write a poor compute shader that expects work-groups to synchronise between themselves.

Yes, that's totally fair, but we need to draw the line *somewhere*. I'd prefer to draw it at the pass boundary (hence, #59), but the group agreed (for now) to draw it at the dispatch boundary for compute work.

> Again that will be a very big DAG, I think you're working under the presumption the DAG will represent a single-frame's worth of work.

Single frame is not my point. I'm trying to say that most resources are read-only, and there isn't an enormous amount of the rest to synchronize access. Besides, as I mentioned in a neighboring thread, we *still* have to track this DAG for validation...

> Also, no matter how good of a portable front-end API you design you will have to implement it in a back-end API, which is not portable.

I disagree. We, being the clients of native APIs, are responsible for using it in a portable way.

**Comment by devshgraphicsprogramming** [comment]
> but the group agreed (for now) to draw it at the dispatch boundary for compute work.

And for graphics work, where do you draw the boundary?
Because dependencies need to be defined between draws (graphics shaders can write to SSBO and storage images too), not between renderpasses/subpasses.

> I disagree. We, being the clients of native APIs, are responsible for using it in a portable way.

You can, with significant effort.

**Comment by devshgraphicsprogramming** [comment]
> we still have to track this DAG for validation...

If by validation you mean making sure the underlying native API object is not deleted before the GPU is done using it, then I don't believe you need a DAG for that.... your standard Garbage Collector enhanced with a reference to a API fence placed after the last use (detected by GC) will do.

**Comment by devshgraphicsprogramming** [comment]
> insert memory barriers

Memory barrier is not a fine-grained memory dependency, also these memory dependencies will not kill the purpose of a render pass on tiling hardware.

Let me give you an example:
You have a tiled deferred renderer that uses a Z-Prepass.

This is your Graphics workload:
Rasterize Z-Prepass subpass
[Does not actually depend on Z-Prepass finishing, can overlap] Rasterize GBuffer subpass
[Does not require any extra dependencies due to per-pixel hardware ordering guarantees] Rasterize albedo decals, environment probes and emissive lighting with blending
[Depends all GBuffer modifications complete + All the compute workloads] Deferred Lighting (this would actually be done in a fragment shader as opposed to compute to make sure you have subpass inputs available to leverage tilers/caches)

Now These could be your INDEPENDENT compute workloads:

Set 0:
[Depends nothing from graphics] Particle System Simulation 

Set 1:
[Depends on Z-Prepass] Min-max Z-Resolve
Set 2:
[Depends on Min-max Z-Resolve] Light-Tile Culling
Set 3:
[Depends on Min-max Z-Resolve] Participating Media Volumetric Lighting


Now you can & should have the memory dependencies between different passes/sets of compute work.

> However, unlike with compute passes, it's too performance heavy to try to synchronize those.

If you cannot achieve this for graphics work, why try it for compute?

> We need to guarantee that the same resource isn't accessed at a time from different places
Why?

**Comment by kvark** [comment]
@devshgraphicsprogramming I think we are losing touch here. I'm saying that it's not feasible to insert any memory barriers *within a render (sub)pass*. Vulkan's support for those is extremely limited, and we weren't able to find a representative use case on why those should be done by us.

> Why? (examples)

To clarify, we need to guarantee that any mutable access to a resource is unique. I.e. you can still read the same resource from different draw calls/stages, but modifying it (without proper synchronization in place) is going to be a source of non-portable behavior that isn't suitable for the Web.

**Comment by devshgraphicsprogramming** [comment]
> I'm saying that it's not feasible to insert any memory barriers within a render (sub)pass. 

You are the ones advocating implicit insertion of synchronisation in-place of the user who should know his data dependencies best. 

In the above example some dependencies must be inserted to get anything approaching the correct result, some can be skipped.
In the above example just the virtue of read-write access to a Z-Buffer would force your theoretical resource tracker to synchronise every dispatch/draw that touches it.

I do acknowledge that somehow you want to grant FBO attachments some special privilege of being exempt from this implicit ordering, but let me point out to you that there exist rendering techniques that forego using classical attachments (A-Buffer) in favour of using UAV's in their place.

> but modifying it (without proper synchronization in place) is going to be a source of non-portable behavior that isn't suitable for the Web.

What about all the programmable blending, compute and order-independent-transparency techniques that depend on read-write access to the same resource both within a dispatch/draw-call as well as across draw-calls and dispatches?

The whole algorithm of Knowles on Order Independent Transparency would no longer become interactive if you disallowed non-deterministic and non-repeatable memory modification.

**Comment by devshgraphicsprogramming** [comment]
The prime example of a thing impossible under your implicit synchronisation proposal and portability constraints is a GPU-created Atomic Linked List.

**Comment by devshgraphicsprogramming** [comment]
>  I think we are losing touch here. 

My overarching point here is that your working group's proposal and concern with keeping mutable access to a resource unique at a renderpass/dispatch(or compute pass) granularity is essentially fighting windmills.

Since it is impossible/impractical to achieve the portable and deterministic behaviour you wish for without stripping the API of much of the features that make the native APIs so enticing.

**Comment by kvark** [comment]
@devshgraphicsprogramming 

> In the above example some dependencies must be inserted to get anything approaching the correct result, some can be skipped.

Sorry, I'm having trouble following your example. Could you explain in a bit more detail what it's demonstrating? If it's to counter my point that we don't want in-render-pass synchronization, where does the example need it?

> In the above example just the virtue of read-write access to a Z-Buffer would force your theoretical resource tracker to synchronise every dispatch/draw that touches it.

Which access exactly do you mean?
  - draws into Z-buffer don't need to be synchronized as the HW ordering guarantees are in place (like you mentioned)
  - there is going to be an UAV barrier between "Min-max Z-Resolve" and the following dispatches, sure. Do you think this is a big problem?

> there exist rendering techniques that forego using classical attachments (A-Buffer) in favour of using UAV's in their place
> What about all the programmable blending, compute and order-independent-transparency techniques that depend on read-write access to the same resource both within a dispatch/draw-call as well as across draw-calls and dispatches?
> The whole algorithm of Knowles on Order Independent Transparency would no longer become interactive if you disallowed non-deterministic and non-repeatable memory modification.

So these appear to be the cases where unsynchronized UAV access is needed across dispatches, right? If so, can you provide some links? If the group missed those cases and we decide that we need to support the efficiently, we may re-consider the compute pass synchronization scope.

> Since it is impossible/impractical to achieve the portable and deterministic behaviour you wish for without stripping the API of much of the features that make the native APIs so enticing.

Metal is a native API that is performant, and it does the synchronization (for basic scenarios, Metal provides means to do that manually as well) on behalf of the user. I hear your skepticism, but so far we haven't got substantial evidence that going this path is wrong. Perhaps, we just need you to support the point with more concrete examples (of things we'd do slow with the current model).

**Comment by devshgraphicsprogramming** [comment]
I will be back with concrete examples, diagrams and explanations.

For the time being I link you to Pyarelal Knowles' body  of OIT work
https://researchbank.rmit.edu.au/eserv/rmit:161520/Knowles.pdf
http://goanna.cs.rmit.edu.au/~pknowles/bma-preprint.pdf

as well as more relevant to UAV synch:
https://www.seas.upenn.edu/~pcozzi/OpenGLInsights/OpenGLInsights-SparseVoxelization.pdf (empty/attachmentless FBO and storage image used in place)

https://www.google.co.uk/url?sa=t&rct=j&q=&esrc=s&source=web&cd=1&ved=2ahUKEwjbgqnP29neAhUBkiwKHX5-DEUQFjAAegQICRAC&url=https%3A%2F%2Fresearch.activision.com%2Feikmo72643%2Fattachments%2Feikmo72643%2Ftech-reports%2F19%2F6%2FPracticalOIT.pdf&usg=AOvVaw066lIeOcyUIXzL9DkVr1ae
(Using CS instead of rasterizer to render particle sprites inbetween transparent OIT geometry)

>  (of things we'd do slow with the current model).

I have a little issue following your current model, mostly because the propositions are split between 3 issues and so on.
It would help if I could refer to a single formal specification of what are the resources you intend to implicitly synchronise, the scope of the implicit synchronisation and the points at which it would occur.
Also a mention (possible implementation) of the synchronisation primitives you would use in at least one of D3D12 or Vulkan.
Furthermore a clear delineation between synchronisation as execution dependencies and memory dependencies.
Lastly a similar specification of scope of your "mutability uniqueness" (as obviously it cannot be ensured at a granularity less than that of a draw/dispatch)
